File: oo-lib-py/tests/test_app.py

import pytest
from app.v1.items import Id, Book, Dvd
from app.v1.catalog import Catalog
from app.v1.member import Member

def setup_sample():
    cat = Catalog()
    cat.add(Book(Id("B1"), "Rust for Humans"))
    cat.add(Book(Id("B2"), "Pythonic Patterns"))
    cat.add(Dvd(Id("D1"), "Taking Flight"))
    return cat

def test_add_and_get():
    cat = setup_sample()
    assert cat.get("B1").title == "Rust for Humans"
    assert cat.get("D1").days_allowed() == 7

def test_duplicate_id_rejected():
    cat = setup_sample()
    with pytest.raises(ValueError):
        cat.add(Book(Id("B1"), "Duplicate"))

def test_member_borrow_and_return():
    cat = setup_sample()
    m = Member("Clark")
    m.borrow("B1")
    assert m.borrowed_ids() == ["B1"]
    m.return_item("B1")
    assert m.borrowed_ids() == []

def test_member_cannot_borrow_twice():
    m = Member("X")
    m.borrow("K1")
    with pytest.raises(ValueError):
        m.borrow("K1")

def test_member_limit_enforced():
    m = Member("X")
    for i in range(5):
        m.borrow(f"K{i}")
    with pytest.raises(ValueError):
        m.borrow("K5")

def test_details_lines():
    cat = setup_sample()
    m = Member("Y")
    m.borrow("B2")
    lines = m.list_details(cat)
    assert len(lines) == 1
    assert "Pythonic Patterns" in lines[0]
    assert "(14 days)" in lines[0]

# Newly added tests by Patrick below

def test_member_cannot_return_unborrowed_id():
    cat = setup_sample()
    m = Member("Adam")
    m.borrow("B1")
    with pytest.raises(ValueError):
        m.return_item("D1")

def test_member_cannot_borrow_items_not_in_catalog():
    cat = setup_sample()
    m = Member("Adam")
    m.borrow("Z8")
    assert m.is_borrowed_list_valid(cat) == False

File: oo-lib-rust/tests/test_app.rs
use oo_lib_rust::v1::member::Member;
use oo_lib_rust::v1::catalog::Catalog;
use oo_lib_rust::v1::items::{Book, Dvd};

fn setup_sample() -> Catalog {
    let mut cat = Catalog::new();
    cat.add(Box::new(Book::new("B1", "Rust for Humans"))).unwrap();
    cat.add(Box::new(Book::new("B2", "Pythonic Patterns"))).unwrap();
    cat.add(Box::new(Dvd::new("D1", "Taking Flight"))).unwrap();
    cat
}

#[test]
fn test_add_and_get() {
    let cat = setup_sample();
    assert_eq!(cat.get("B1").unwrap().get_title(), "Rust for Humans");
    assert_eq!(cat.get("D1").unwrap().days_allowed(), 7);
}

#[test]
fn test_duplicate_id_rejected() {
    let mut cat = setup_sample();
    let result = cat.add(Box::new(Book::new("B1", "Duplicate")));
    assert!(result.is_err());
}

#[test]
fn test_member_borrow_and_return() {
    let mut m = Member::new("Clark");
    m.borrow("B1").unwrap();
    let mut vec = vec!["B1".to_string()];
    assert_eq!(m.borrowed_ids(), &vec);
    
    m.return_item("B1").unwrap();
    vec.pop(); // []
    assert_eq!(m.borrowed_ids(), &vec); // [] == [] 
}

#[test]
fn test_member_cannot_borrow_twice() {
    let mut m = Member::new("Adam");
    m.borrow("K1").unwrap();
    let result = m.borrow("K1");
    assert!(result.is_err());
}

#[test]
fn test_member_limit_enforced() {
    let mut m = Member::new("Adam");
    for i in 0..5 {
        m.borrow(&format!("K{}", i)).unwrap();
    }
    let result = m.borrow("K5");
    assert!(result.is_err());
}

#[test]
fn test_details_lines() {
    let cat = setup_sample();
    let mut m = Member::new("Adam");
    m.borrow("B2").unwrap();
    let lines = cat.details_for(&m.borrowed_ids());
    assert_eq!(lines.len(), 1);
    assert_eq!(lines[0].0, "B2".to_string());
    assert_eq!(lines[0].1, "Pythonic Patterns".to_string());
    assert_eq!(lines[0].2, 14);
}

#[test]
fn test_member_cannot_return_unborrowed_id() {
    let mut m = Member::new("Adam");
    m.borrow("M1").unwrap();
    let result = m.return_item("Z10");
    assert!(result.is_err());
}

#[test]
fn test_member_cannot_borrow_items_not_in_catalog() {
    let cat = setup_sample();
    let mut m = Member::new("Zeke");
    m.borrow("A24").unwrap();
    assert_eq!(m.is_borrowed_list_valid(&cat), false);
}
